### DISCUSSION TITLE
Bulk registering from base class and resolving all subclasses

### DISCUSSION THREAD

**Discussion by evan-williams** [dadhi/DryIoc#556] [Q&A] [answered]
Hi - apologies if this is obvious from the docs but I've read them multiple times and tried all sorts of things but can't get this to work and I'm sure it's me and not Dryloc i.e. this would be a supported scenario.

Say I have a base class called BaseEvent and multiple classes that inherit from that.
I'm trying to register like this:

```
// Specific, register-once, service-implementation types here
container.Register<IRepository,RepositorySqlServer>(Reuse.Scoped);
....

container.RegisterMany(new[] { typeof(BaseEvent).Assembly }, 
                type => type.IsSubclassOf(typeof(BaseEvent)),
                reuse:Reuse.Transient,
                ifAlreadyRegistered:IfAlreadyRegistered.AppendNewImplementation);

// Catch-all
container.RegisterMany(
                Assembly.GetAssembly(typeof(IRepository))?.GetTypes(), 
                reuse:Reuse.Transient, 
                serviceTypeCondition: (type => (type.IsInterface || type.IsClass) && !type.IsAbstract));
```

but when I try resolve using 
`resolver.ResolveMany<BaseEvent>()`
or
`resolver.Resolve<IEnumerable<BaseEvent>>()`

I keep getting just the base class itself and not all the implementations.

Please let me know what I'm doing wrong. Thanks so much.

**Comment by dadhi** [comment]
@evan-williams 
Hi. First, the second call to `RegisterMany` may override some registrations from the first one, right?
Second, could add some implementation classes into the example and make it complete?
Otherwise, it may be too many factors influencing the issue and without the context I cannot guess.

**Reply by evan-williams** [reply]
Hi dadhi - thanks for getting back to me so quickly.

I hadn't thought of that - not sure why I thought the default for `ifAlreadyRegistered `would be `Keep `(i.e. ignore if already registered) for the behaviour of the catch-all. Maybe I'm used to that coming from Windsor.

I knocked up a small demo (attached) which seems to work as expected so probably you were right about the catch-all overwriting my explicit registration from before.

I managed to get my desired behavior with something like this (where `GetSubClasses` was explicitly excluding the base/abstract classes from the output):

`container.RegisterMany(GetSubClassesOf<BaseEvent>(), reuse:Reuse.Transient, ifAlreadyRegistered:IfAlreadyRegistered.AppendNewImplementation);`

[DrylocTest.zip](https://github.com/dadhi/DryIoc/files/10725707/DrylocTest.zip)

Anyway thanks for your help

**Reply by dadhi** [reply, marked as answer]
@evan-williams Glad to hear that you've solved it.
